### PR TITLE
Explicit-comms: preserve partition IDs

### PR DESCRIPTION
**Pull request by madsbk**
`shuffle_task()` now returns a dict mapping partition IDs to dataframes`

Fixes https://github.com/rapidsai/dask-cuda/issues/1239

### CODE REVIEW

**Comment by pentschev** [comment]
Could we add the sample from #1239 or some variation of it as a test?

**Review by rjzamora — APPROVED** [review]
Thanks @madsbk !

File: dask_cuda/tests/test_explicit_comms.py [hunk 1]
```diff
@@ -123,33 +126,46 @@ def _test_dataframe_shuffle(backend, protocol, n_workers):
...
 
                             assert ddf.npartitions == output_nparts
 
-                            # Check that each partition hashes to the same value
-                            result = ddf.map_partitions(
-                                check_partitions, output_nparts
-                            ).compute()
-                            assert all(result.to_list())
+                            if _partitions:
+                                # If "_partitions" is the hash key, we except all but
```
Comment (rjzamora) [inline dask_cuda/tests/test_explicit_comms.py:137]:
```suggestion
                                # If "_partitions" is the hash key, we expect all but
```

File: dask_cuda/tests/test_explicit_comms.py [hunk 2]
```diff
@@ -123,33 +126,46 @@ def _test_dataframe_shuffle(backend, protocol, n_workers):
...
+                                # the first partition to be empty
+                                assert_eq(ddf.partitions[0].compute(), df)
+                                assert all(
+                                    len(ddf.partitions[i].compute()) == 0
+                                    for i in range(1, ddf.npartitions)
+                                )
+                            else:
+
```
Comment (rjzamora) [inline dask_cuda/tests/test_explicit_comms.py:145]:
```suggestion
```
Super-nit

File: dask_cuda/tests/test_explicit_comms.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pentschev) [reply dask_cuda/tests/test_explicit_comms.py:137]: Is this supposed to read "accept", or am I misunderstanding the meaning of the sentence?

**Review by pentschev — COMMENTED** [review]
LGTM, just a minor question/suggestion on one of your comments.

File: dask_cuda/tests/test_explicit_comms.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjzamora) [reply dask_cuda/tests/test_explicit_comms.py]: WOW, my mistake!

**Review by pentschev — APPROVED** [review]
Thanks @madsbk and @rjzamora .

**Pull request merged**